model/src/app_id.rs

```diff
@@ -0,0 +1,99 @@
+use serde::{de, Deserializer, Serializer};
+use serde_with::serde::{Deserialize, Serialize};
+use std::{
+    fmt::{self, Debug, Formatter},
+    str::FromStr,
+};
+
+#[derive(Clone, Copy, Default, Eq, Hash, PartialEq)]
+pub struct AppId(pub [u8; 32]);
+
+impl Debug for AppId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+impl FromStr for AppId {
+    type Err = hex::FromHexError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut bytes = [0u8; 32];
+        hex::decode_to_slice(s.strip_prefix("0x").unwrap_or(s), &mut bytes)?;
+        Ok(Self(bytes))
+    }
+}
+
+impl Serialize for AppId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut bytes = [0u8; 2 + 32 * 2];
+        bytes[..2].copy_from_slice(b"0x");
+        // Can only fail if the buffer size does not match but we know it is correct.
+        hex::encode_to_slice(self.0, &mut bytes[2..]).unwrap();
+        // Hex encoding is always valid utf8.
+        let s = std::str::from_utf8(&bytes).unwrap();
+        serializer.serialize_str(s)
+    }
+}
+
+impl<'de> Deserialize<'de> for AppId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
```

Comment:
ubernit: We can use a `Cow<str>` to potentially avoid additional allocations.
```suggestion
        let s = Cow::<str>::deserialize(deserializer)?;
```